### PR TITLE
New external plugins added

### PR DESCRIPTION
**Pull request by yeuk0**
pybuilder_pycharm_workspace for PyCharm workspace creation
pybuilder_archetype_base and pybuilder_archetype_api for project structure generation

### CODE REVIEW

**Comment by arcivanov** [comment]
Just curious, how's `pybuilder_pycharm_workspace` different from `https://github.com/pybuilder/pybuilder/blob/master/src/main/python/pybuilder/plugins/python/pycharm_plugin.py`?

**Comment by yeuk0** [comment]
As I saw -correct me if I'm wrong- that task only tells PyCharm which one is the source directory (unit tests directories too?) and which directories it should ignore.

`pybuilder_pycharm_workspace` plugin sets that metadata too furthermore it adds:

- `.idea` directory ignoring
- `target/dist` directory ignoring only (logs from PyBuilder visible)
- Creates PyCharm Python interpreter configuration associated with project's virtual environment (`venv/` by default) and sets it as default for the project
- Creates some PyCharm running profiles ready to go that launches PyBuilder preconfigured tasks and unit tests with just one click. These profiles look like the following:
imagen

Hoping it's clearer now. Let me know any other doubt you have related with this one or any of the other plugins added.

**Comment by arcivanov** [comment]
Oh wow, that's awesome! Thanks!

As far as venv, please check out 0.12.x (master branch), as it manages venvs automagically:
- https://github.com/pybuilder/pybuilder/blob/master/src/main/python/pybuilder/plugins/python/core_plugin.py#L61
- https://github.com/pybuilder/pybuilder/blob/master/src/main/python/pybuilder/reactor.py#L511

**Comment by yeuk0** [comment]
Looks cool. I built these plugins using 0.12 dev versions and didn't notice this venvs. However I have some doubts about this approach I hope you don't mind to clarify:

1. I have two new venv: build and tests. I guess the second one is used by PyBuilder for executing the tests etc. But the first one? This should be the one used during the launching/developing of the project? I'm confused at this point, sorry.
2. In case the answer for the previous question is 'yes', why the venv name is the Python version used in PyBuilder launching? (`cpython-3.7.3.final.0` in my case). I mean, why including this instead of having venv inside `target/venv/build` directly?
If this is the expected behaviour, the automatic configuration of Python interpreter for PyCharm will be a pain in the ass literally haha.

tl;dr Are you supposed to launch `pyb prepare` before starting developing (for example)? I mean, taking in mind what I usually do (create venvs before opening PyCharm), I guess this should be the correct approach now, right?

**Comment by arcivanov** [comment]
@yeuk0 

This is going to be in release notes and docs, but briefly:

1. `Build` venv is the environment with `build` and `install` dependencies. This is where unit tests are run and most of the other tooling. `Test` venv is where the integration tests run and it only contains `install` dependencies, as it ensures maximum purity of the venv that is not polluted with build dependencies that may affect how things run.
2. PyBuilder has to work with multiple types and versions of Python (CPython, PyPy etc). You may, within the same source tree switch between CPython 2.7, 3.5 and 3.8. You naturally can't cross-use the venvs between Python versions because dependencies installed (even if they are the same versions) could be different code bases/compiled binaries.
3. In the future PyBuilder may even manage Python versions via pyenv/chocolatey via separate plugin and allow concurrent building testing in multiple python versions.

**Comment by arcivanov** [comment]
You're not supposed to run `pyb prepare`, as it will be run automatically as part of the build stage dependencies. The venvs will be created/updated and maintained automatically based on the build descriptor.

In version 0.11 you had to do `pyb install_dependencies`. Now you don't have to maintain the venv at all.

I would recommend reviewing how Python builds and tests itself for more info:
https://travis-ci.org/github/pybuilder/pybuilder/jobs/668570820#L275
https://travis-ci.org/github/pybuilder/pybuilder/jobs/668570821#L279
